### DISCUSSION TITLE
Assigned errors must be positive, in hypotest

### DISCUSSION THREAD

**Discussion by ssaha1234** [scikit-hep/pyhf#1988] [Q&A] [answered]
Dear pyhf members, 
        I am running a simple hypotest for producing limit plots for my analysis. I am using pyfh 0.6.3, following is a short snippet of the code

```python
pyhf.set_backend("JAX", pyhf.optimize.minuit_optimizer(maxiter=int(1e9), tolerance=10000000))
wspace = pyhf.Workspace(json.load(open("LH_CLs_{}.json".format(int(sys.argv[2])))))
model = wspace.model()
data = wspace.data(model)

result = pyhf.infer.hypotest(float(sys.argv[3]), data, model, return_expected_set=True)
retval = [result, float(sys.argv[2]), float(sys.argv[3])]
pickle.dump(retval, open("CLs_result_{}.pkl".format(int(sys.argv[1])), "wb"))
```

I am getting the following user warning - 
`Assigned errors must be positive. Non-positive values are replaced by a heuristic.`

after which the job goes into `hold` status on Condor. I do not have any negative values in the json file, I do not understand the source of this warning, is there any hint for solving this issue. I do see something related to the error here - https://github.com/scikit-hep/pyhf/pull/1919 but I am not sure I understand it completely. Any help would be welcome ! 

P.S. - I released the jobs on Condor, so I am hoping I will get results soon. 

Thank you ! 
Best, 
Shreya

**Comment by kratsg** [comment]
Why is your tolerance very large? This seems wrong.

**Reply by ssaha1234** [reply]
Hi @kratsg, 
     It is because the fit was not converging with a smaller tolerance :( Ofcourse, I can reduce it a little and check, what is the accepted value of tolerance, in general ?
Thanks !
Best, 
Shreya

**Reply by alexander-held** [reply]
The default of `0.1` is good, going lower to `1e-3` or `1e-4` can be useful in some cases. If it does not converge with these settings, there is likely a problem in your model.

**Reply by matthewfeickert** [reply]
Yes, that is indicative of a problem with the model.

**Reply by ssaha1234** [reply]
Hi @alexander-held  and @matthewfeickert, 
     Okay I understand, I will try with a smaller tolerance and check if the fit converges or not. 
Thanks ! 
Best, 
Shreya

**Comment by matthewfeickert** [comment, marked as answer]
> I am getting the following user warning - `Assigned errors must be positive. Non-positive values are replaced by a heuristic.`
> 
> after which the job goes into `hold` status on Condor. I do not have any negative values in the json file, I do not understand the source of this warning, is there any hint for solving this issue.

PR #1919 resolved Issue #1918 and will go into `pyhf` `v0.7.0`. To avoid getting this warning during rumtime now with `pyhf` `v0.6.3` just use `iminuit<2.12.2`. So assuming you're installing `pyhf` `v0.6.3` with the `minuit` extra:

```
# requirements.txt
pyhf[minuit]==0.6.3
iminuit<2.12.2
```

```console
$ python -m pip install -r requirements.txt
```

**Reply by ssaha1234** [reply]
Hi @matthewfeickert, 
    Thanks for the quick reply, I will check the version of `iminuit` I am using currently, and make sure it is below 2.12.2. Thanks for linking #1918, I understand it a bit better now ! 
Thanks. 
Best, 
Shreya
